### PR TITLE
ref(signals): Consolidate sigint/sigterm usage under atexit

### PR DESCRIPTION
**Pull request by BYK**
Instead of sprinkling incomplete signal handlers over the codebase
prefer using `atexit.register` and trigger an exit upon receiving
`SIGTERM` as it normally bypasses the `atexit` callbacks.

The new code should behave identical to the old one, but more complete
as in it now handles more exit triggers instead of just `SIGINT`.

This is a follow up to #15934.

### CODE REVIEW

**Review by untitaker — DISMISSED** [review]

File: src/sentry/runner/__init__.py [hunk 1]
```diff
@@ -161,4 +162,8 @@ def call_command(name, obj=None, **kwargs):
 
 
 def main():
+    def sigHandler(signo, frame):
```
Comment (untitaker) [inline src/sentry/runner/__init__.py]: can we make this snake_case. Otherwise lgtm

**Review by tkaemming — CHANGES_REQUESTED** [review]
I am ~nearly~ certain this will not work as expected.

Using the example of the batching consumer and SIGTERM:

1. The batching consumer is started on the main thread via `run`.
2. The running process receives SIGTERM.
3. The interpreter suspends execution of the main thread to invoke the signal handler.
4. The signal handler requests the process exit via `sys.exit`. The `atexit` handlers are called (note that we are still in the execution context of the signal handler, *not* the main thread), which requests that the batching consumer begin its shutdown process via the non-blocking `signal_shutdown` method.
5. The main thread resumes, immediately throws the `SystemExit` exception at its current frame. That unwinds the call stack of the consumer all the way to the top, and the interpreter crashes out. The consumer never exits cleanly.

See this example: https://gist.github.com/tkaemming/c7885d619213c4282a1ee5caa4bc36cd

If you wrap the `consumer.run()` call like this:

```python
try:
    consumer.run()
except SystemExit as e:
    logger.exception(e)
```

what you will see is this instead:

```
MainThread DEBUG Starting consumer run loop...
MainThread DEBUG Received signal: 2
MainThread ERROR 0
Traceback (most recent call last):
  File "test.py", line 40, in <module>
    consumer.run()
  File "test.py", line 21, in run
    while not self.shutdown:
  File "test.py", line 33, in handler
    sys.exit(0)
SystemExit: 0
MainThread DEBUG Main thread exited cleanly.
Dummy-1 DEBUG Consumer shutdown requested...
```

As implemented, this also doesn't handle SIGINT, which will just throw a `KeyboardInterrupt` and behave similarly (though that one will actually trigger the except hook and we might see it logged to Sentry.)

**Comment by BYK** [comment]
Turns out a better way to break cleanly is simply handling the `SystemExit` exception in this case:
```py
    def run(self):
        logger.debug("Starting consumer run loop...")
        try:
            while not self.shutdown:
                pass
        except SystemExit:
            logger.debug("System shutdown detected...")
        logger.debug("Consumer shut down cleanly.")
```

produces the following output:
```
MainThread DEBUG Starting consumer run loop...
^CMainThread DEBUG Received signal: 2
MainThread DEBUG System shutdown detected...
MainThread DEBUG Consumer shut down cleanly.
MainThread DEBUG Main thread exited cleanly.
Dummy-1 DEBUG Consumer shutdown requested...
```

At that point, `signal_shutdown` and `atexit` handler both become obsolete. @tkaemming you okay going with this approach? By doing `except (SystemExit, KeyboardInterrupt)` we can remove the need for `SIGINT` handling too (tested and verified).

**Comment by BYK** [comment]
Another reason to do this is signal handlers being unique per app so it is easy to step on others' toes when registering signal handlers like we do currently in the code.

**Comment by tkaemming** [comment]
> Turns out a better way to break cleanly is simply handling the `SystemExit` exception in this case:
> 
> ```python
>     def run(self):
>         logger.debug("Starting consumer run loop...")
>         try:
>             while not self.shutdown:
>                 pass
>         except SystemExit:
>             logger.debug("System shutdown detected...")
>         logger.debug("Consumer shut down cleanly.")
> ```
> 

No, it's not. This still interrupts the work of the consumer and discards whatever batch is in process. This can lead to reprocessed message batches, duplicated writes, or other issues depending on the consumer implementation. It's not accurate or reasonable to describe this as a "clean shutdown". The current pattern of asynchronously handling signals is not an uncommon or unique one. I don't see any advantage to making this a synchronous process.

> Another reason to do this is signal handlers being unique per app so it is easy to step on others' toes when registering signal handlers like we do currently in the code.

This sounds more like you have an ideological issue with people mutating global state, which is not totally unreasonable. We haven't had any specific issues with this when dealing with signal handler registration, which I'm going to guess is probably because very few pieces of code need to actually handle signals. That doesn't mean that it won't ever happen, but until now this has never been a problem (and in my opinion, isn't a problem now either.)

These consumers are always blocking the main thread, so after their signals are registered, I think it's fair to hold the implementer of the consumer responsible for understanding for the code that it executes. If that code for some reason attaches signal handlers that clobber the existing ones, that sounds to me like either the implementer of the consumer didn't understand the code that they were calling, or the code that they were calling is of dubious quality (either the author didn't understand signals, or the architecture was bad.)

**Comment by BYK** [comment]
>No, it's not. This still interrupts the work of the consumer and discards whatever batch is in process. This can lead to reprocessed message batches, duplicated writes, or other issues depending on the consumer implementation. It's not accurate or reasonable to describe this as a "clean shutdown". The current pattern of asynchronously handling signals is not an uncommon or unique one. I don't see any advantage to making this a synchronous process.

Yes, sorry my bad. Upon looking deeper, the approach I took in this PR is fundamentally different, potentially causing a half-executed loop body, which certainly is not desirable or wanted and the direct signal approach definitely prevents this issue.

>This seems more like an issue with mutating global state than with signal registration itself (and I generally think avoiding mutating global state is a good thing, to be clear), but we haven't had any specific issues with this when dealing with signal handler registration, which I'm going to guess is probably because very few pieces of code need to actually handle signals. That doesn't mean that it won't ever happen, but until now this has never been a problem (and in my opinion, isn't a problem now either.)

Not necessarily. I'm not a fan of non-explicit global state mutations but my worry here was us having an incoherent pattern regarding this termination handling code and the leaky abstraction that is leaving the clean shutdown management job to the consumer of the class. That said what I was missing is all these are blocking, infinite while loops on the main thread so there's little to no risk of signal handler collusion. Moreover, both of these code paths seem to be used by dedicated runner processes, which makes the approach totally reasonable, just not very safe or scalable in case someone tried to reuse these classes in a different context as you have stated.

Based on your feedback (and finally grasping the entirety of how all this works, I think this PR and the idea it is pursuing is not needed at all until some later time (if ever) so closing it out.

This was a very educational journey so thank you very much for all the comments and code snippets helping me understand the situation better.

**Comment by tkaemming** [comment]
> Not necessarily. I'm not a fan of non-explicit global state mutations but my worry here was us having an incoherent pattern regarding this termination handling code and the leaky abstraction that is leaving the clean shutdown management job to the consumer of the class.

Yeah, I agree. I think it's going to be very difficult to centralize this without defining some sort of abstract interface for entrypoints that allows them to define but not install their own signal handlers, which would then be installed by the runner. This obviously wouldn't be able to stop anybody from directly using `signal.signal` to clobber existing ones, but at least would define a pattern for where signal handlers should be registered. Without consolidating this so that signal handlers are *always* registered in a single location (at least by convention), I think we'd just be moving the problem around instead of really solving it.

> Moreover, both of these code paths seem to be used by dedicated runner processes, which makes the approach totally reasonable, just not very safe or scalable in case someone tried to reuse these classes in a different context as you have stated.

I think making the implicit coupling between these consumers and the runner more obvious would be a good step in the right direction here, since it would discourage reuse outside of that context. That said, I don't have any good ideas on how to do that in a way that seems to balance the return on investment vs. the implementation complexity — but I've also spent  next to no time thinking about it, so maybe you can come up with something.

> This was a very educational journey so thank you very much for all the comments and code snippets helping me understand the situation better.

For sure. I think the fact that it took so much effort on both of our sides to come to an understanding here is evidence of some design issues and I think gives some credibility to your overall point, regardless of the implementation details.

**Pull request closed without merging**